File: Neetcode Questions/Stack/daily-temperatures.cpp

// https://leetcode.com/problems/daily-temperatures/description/

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        
        vector<int> ans;
        if(temperatures.size() == 0) return ans;
        reverse(temperatures.begin(), temperatures.end());
        stack<int> st;
        ans.push_back(0);
        st.push(0);
        
        for(int i = 1; i < temperatures.size(); i++){
            while(!st.empty() && temperatures[i] >= temperatures[st.top()]) st.pop();
            if(!st.empty()) ans.push_back(i - st.top());
            else ans.push_back(0);
            st.push(i);
        }
        
        reverse(ans.begin(), ans.end());
        
        return ans;
    }
};

File: Neetcode-150-js/Stack/daily-temperatures.js

// https://leetcode.com/problems/daily-temperatures/description/

/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
 
 Array.prototype.empty = function(){
   return this.length === 0;
 }
 
 Array.prototype.back = function(){
   return this[this.length-1];
 }
 
var dailyTemperatures = function(temperatures) {

    let len = temperatures.length;
    const answer = Array.from({length : len}, () => 0);
    if(len === 0) return answer;

    temperatures = temperatures.reverse();
    const stack = [0];
    
    for(let i = 1; i < len; i++){
      while(!stack.empty() && temperatures[i] >= temperatures[stack.back()]) stack.pop();
      
      if(stack.empty()) answer[i] = 0;
      else answer[i] = i - stack.back();
      
      stack.push(i);
    }
    
    return answer.reverse();
};